README.md
# [@root/request](https://git.rootprojects.org/root/request.js) | a [Root](https://rootprojects.org) project

> Minimalist HTTP client

A lightweight alternative to (and 80/20 drop-in replacement for) request.

Has the 20% of features that 80%+ of people need, in about 500 LoC.

Written from scratch, with zero-dependencies.

## Super simple to use

@root/request is designed to be a drop-in replacement for request. It also supports Promises and async/await by default.

```bash
npm install --save @root/request

# or npm install git+ssh://git@git.therootcompany.com/request.js
```

```js
var request = require('@root/request');
request('http://www.google.com', function (error, response, body) {
    console.log('error:', error); // Print the error if one occurred
    console.log('statusCode:', response && response.statusCode); // Print the response status code if a response was received
    console.log('body:', body); // Print the HTML for the Google homepage.
});
```

**Using Promises**

```js
var request = require('@root/request');

request('http://www.google.com')
    .then(function (response) {
        console.log('statusCode:', response.statusCode); // Print the response status code if a response was received
        console.log('body:', response.body); // Print the HTML for the Google homepage.
    })
    .catch(function (error) {
        console.log('error:', error); // Print the error if one occurred
    });
```

**Streaming**

In order to keep this library lightweight, performant, and keep the code easy to
read, the streaming behavior is **_slightly different_** from that of
`request.js`.

```diff
-var request = require('request');
+var request = require('@root/request');
 
-var stream = request({ url, headers });
+var stream = await request({ url, headers });

 let attachment = await new MailgunAPI.Attachment({
   data: stream
 })
```

Example:

```js
var request = require('@root/request');

var resp = await request({
    url: 'http://www.google.com',
    stream: true // true | 'filename.ext' | stream.Writable
});

// 'resp' itself is a ReadableStream
resp.on('data', function () {
    // got some data
});

resp.on('end', function () {
    // the data has ended
});

// 'resp.stream' is a Promise that is resolved when the read stream is destroyed
await resp.stream; // returns `undefined`
console.log('Done');
```

The difference is that we don't add an extra layer of stream abstraction.
You must use the response from await, a Promise, or the callback.

You can also give a file path:

```js
request({
    url: 'http://www.google.com',
    stream: '/tmp/google-index.html'
});
```

Which is equivalent to passing a write stream for the file:

```js
request({
    url: 'http://www.google.com',
    stream: fs.createWriteStream('/tmp/google-index.html')
});
```

## Table of contents

-   [Forms](#forms)
-   [HTTP Authentication](#http-authentication)
-   [Custom HTTP Headers](#custom-http-headers)
-   [Unix Domain Sockets](#unix-domain-sockets)
-   [**All Available Options**](#requestoptions-callback)

## Forms

`urequest` supports `application/x-www-form-urlencoded` and `multipart/form-data` form uploads.

<!-- For `multipart/related` refer to the `multipart` API. -->

#### application/x-www-form-urlencoded (URL-Encoded Forms)

URL-encoded forms are simple.

```js
request.post('http://service.com/upload', { form: { key: 'value' } });
// or
request.post(
    { url: 'http://service.com/upload', form: { key: 'value' } },
    function (err, httpResponse, body) {
        /* ... */
    }
);
```

<!--
// or
request.post('http://service.com/upload').form({key:'value'})
-->

#### multipart/form-data (Multipart Form Uploads)

For `multipart/form-data` we use the [form-data](https://github.com/form-data/form-data/tree/v2.5.1) library by [@felixge](https://github.com/felixge). For the most cases, you can pass your upload form data via the `formData` option.

To use `form-data`, you must install it separately:

```bash
npm install --save form-data@2.x
```

```js
var formData = {
    // Pass a simple key-value pair
    my_field: 'my_value',
    // Pass data via Buffers
    my_buffer: Buffer.from([1, 2, 3]),
    // Pass data via Streams
    my_file: fs.createReadStream(__dirname + '/unicycle.jpg'),
    // Pass multiple values /w an Array
    attachments: [
        fs.createReadStream(__dirname + '/attachment1.jpg'),
        fs.createReadStream(__dirname + '/attachment2.jpg')
    ],
    // Pass optional meta-data with an 'options' object with style: {value: DATA, options: OPTIONS}
    // Use case: for some types of streams, you'll need to provide "file"-related information manually.
    // See the `form-data` README for more information about options: https://github.com/form-data/form-data
    custom_file: {
        value: fs.createReadStream('/dev/urandom'),
        options: {
            filename: 'topsecret.jpg',
            contentType: 'image/jpeg'
        }
    }
};
request.post(
    { url: 'http://service.com/upload', formData: formData },
    function optionalCallback(err, httpResponse, body) {
        if (err) {
            return console.error('upload failed:', err);
        }
        console.log('Upload successful!  Server responded with:', body);
    }
);
```

<!--

For advanced cases, you can access the form-data object itself via `r.form()`. This can be modified until the request is fired on the next cycle of the event-loop. (Note that this calling `form()` will clear the currently set form data for that request.)

```js
// NOTE: Advanced use-case, for normal use see 'formData' usage above
var r = request.post('http://service.com/upload', function optionalCallback(err, httpResponse, body) {...})
var form = r.form();
form.append('my_field', 'my_value');
form.append('my_buffer', Buffer.from([1, 2, 3]));
form.append('custom_file', fs.createReadStream(__dirname + '/unicycle.jpg'), {filename: 'unicycle.jpg'});
```
-->

See the [form-data README](https://github.com/form-data/form-data) for more information & examples.

---

## HTTP Authentication

<!--
request.get('http://some.server.com/').auth('username', 'password', false);
// or
request.get('http://some.server.com/').auth(null, null, true, 'bearerToken');
// or
-->

```js
request.get('http://some.server.com/', {
    auth: {
        user: 'username',
        pass: 'password',
        sendImmediately: false
    }
});
// or
request.get('http://some.server.com/', {
    auth: {
        bearer: 'bearerToken'
    }
});
```

If passed as an option, `auth` should be a hash containing values:

-   `user` || `username`
-   `pass` || `password`
-   `bearer` (optional)

<!--
- `sendImmediately` (optional)

The method form takes parameters
`auth(username, password, sendImmediately, bearer)`.

`sendImmediately` defaults to `true`, which causes a basic or bearer
authentication header to be sent. If `sendImmediately` is `false`, then
`request` will retry with a proper authentication header after receiving a
`401` response from the server (which must contain a `WWW-Authenticate` header
indicating the required authentication method).
-->

Note that you can also specify basic authentication using the URL itself, as
detailed in [RFC 1738](http://www.ietf.org/rfc/rfc1738.txt). Simply pass the
`user:password` before the host with an `@` sign:

```js
var username = 'username',
    password = 'password',
    url = 'http://' + username + ':' + password + '@some.server.com';

request({ url: url }, function (error, response, body) {
    // Do more stuff with 'body' here
});
```

<!--
Digest authentication is supported, but it only works with `sendImmediately`
set to `false`; otherwise `request` will send basic authentication on the
initial request, which will probably cause the request to fail.
-->

Bearer authentication is supported, and is activated when the `bearer` value is
available. The value may be either a `String` or a `Function` returning a
`String`. Using a function to supply the bearer token is particularly useful if
used in conjunction with `defaults` to allow a single function to supply the
last known token at the time of sending a request, or to compute one on the fly.

[back to top](#table-of-contents)

---

## Custom HTTP Headers

HTTP Headers, such as `User-Agent`, can be set in the `options` object.
In the example below, we call the github API to find out the number
of stars and forks for the request repository. This requires a
custom `User-Agent` header as well as https.

```js
var request = require('request');

var options = {
    url: 'https://api.github.com/repos/request/request',
    headers: {
        'User-Agent': 'request'
    }
};

function callback(error, response, body) {
    if (!error && response.statusCode == 200) {
        var info = JSON.parse(body);
        console.log(info.stargazers_count + ' Stars');
        console.log(info.forks_count + ' Forks');
    }
}

request(options, callback);
```

[back to top](#table-of-contents)

---

## UNIX Domain Sockets

`urequest` supports making requests to [UNIX Domain Sockets](https://en.wikipedia.org/wiki/Unix_domain_socket). To make one, use the following URL scheme:

```js
/* Pattern */ 'http://unix:SOCKET:PATH';
/* Example */ request.get(
    'http://unix:/absolute/path/to/unix.socket:/request/path'
);
```

Note: The `SOCKET` path is assumed to be absolute to the root of the host file system.

[back to top](#table-of-contents)

---

## request(options, callback)

The first argument can be either a `url` or an `options` object. The only required option is `uri`; all others are optional.

-   `uri` || `url` - fully qualified uri or a parsed url object from `url.parse()`
-   `method` - http method (default: `"GET"`)
-   `headers` - http headers (default: `{}`)

<!-- TODO
- `baseUrl` - fully qualified uri string used as the base url. Most useful with `request.defaults`, for example when you want to do many requests to the same domain. If `baseUrl` is `https://example.com/api/`, then requesting `/end/point?test=true` will fetch `https://example.com/api/end/point?test=true`. When `baseUrl` is given, `uri` must also be a string.
-->

---

-   `body` - entity body for PATCH, POST and PUT requests. Must be a `Buffer`, `String` or `ReadStream`. If `json` is `true`, then `body` must be a JSON-serializable object.
-   `json` - sets `body` to JSON representation of value and adds `Content-type: application/json` header. Additionally, parses the response body as JSON.

<!-- TODO
- `form` - when passed an object or a querystring, this sets `body` to a querystring representation of value, and adds `Content-type: application/x-www-form-urlencoded` header. When passed no options, a `FormData` instance is returned (and is piped to request). See "Forms" section above.
- `formData` - data to pass for a `multipart/form-data` request. See
  [Forms](#forms) section above.
- `multipart` - array of objects which contain their own headers and `body`
  attributes. Sends a `multipart/related` request. See [Forms](#forms) section
  above.
  - Alternatively you can pass in an object `{chunked: false, data: []}` where
    `chunked` is used to specify whether the request is sent in
    [chunked transfer encoding](https://en.wikipedia.org/wiki/Chunked_transfer_encoding)
    In non-chunked requests, data items with body streams are not allowed.
- `preambleCRLF` - append a newline/CRLF before the boundary of your `multipart/form-data` request.
- `postambleCRLF` - append a newline/CRLF at the end of the boundary of your `multipart/form-data` request.
- `jsonReviver` - a [reviver function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse) that will be passed to `JSON.parse()` when parsing a JSON response body.
- `jsonReplacer` - a [replacer function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) that will be passed to `JSON.stringify()` when stringifying a JSON request body.
-->

---

-   `followRedirect` - follow HTTP 3xx responses as redirects (default: `true`). This property can also be implemented as function which gets `response` object as a single argument and should return `true` if redirects should continue or `false` otherwise.
-   `followAllRedirects` - follow non-GET HTTP 3xx responses as redirects (default: `false`)
-   `followOriginalHttpMethod` - by default we redirect to HTTP method GET. you can enable this property to redirect to the original HTTP method (default: `false`)
-   `maxRedirects` - the maximum number of redirects to follow (default: `10`)
-   `removeRefererHeader` - removes the referer header when a redirect happens (default: `false`). **Note:** if true, referer header set in the initial request is preserved during redirect chain.

---

-   `encoding` - encoding to be used on `setEncoding` of response data. If `null`, the `body` is returned as a `Buffer`. Anything else **(including the default value of `undefined`)** will be passed as the [encoding](http://nodejs.org/api/buffer.html#buffer_buffer) parameter to `toString()` (meaning this is effectively `utf8` by default). (**Note:** if you expect binary data, you should set `encoding: null`.)

<!-- TODO

CHANGELOG.md
# CHANGELOG

## v1.8.0

-   add `resp.ok` - same as WHATWG fetch `resp.ok = (resp.statusCode >= 200 && resp.statusCode < 300)`
-   add `resp.stream.body()` to populate `resp.body` rather than (or perhaps in addition to) continuing to stream (useful for error handling)


index.js
'use strict';

var http = require('http');
var https = require('https');
var url = require('url');
var os = require('os');
var pkg = require('./package.json');
var fs = require('fs'); // only for streams

function debug() {
    if (module.exports.debug) {
        console.log.apply(console, arguments);
    }
}

function mergeOrDelete(defaults, updates) {
    Object.keys(defaults).forEach(function (key) {
        if (!(key in updates)) {
            updates[key] = defaults[key];
            return;
        }

        // neither accept the prior default nor define an explicit value
        // CRDT probs...
        if ('undefined' === typeof updates[key]) {
            delete updates[key];
        } else if (
            'object' === typeof defaults[key] &&
            'object' === typeof updates[key]
        ) {
            updates[key] = mergeOrDelete(defaults[key], updates[key]);
        }
    });

    return updates;
}

// retrieves an existing header, case-sensitive
function getHeaderName(reqOpts, header) {
    var headerNames = {};
    Object.keys(reqOpts.headers).forEach(function (casedName) {
        headerNames[casedName.toLowerCase()] = casedName;
    });
    // returns the key, which in erroneous cases could be an empty string
    return headerNames[header.toLowerCase()];
}
// returns whether or not a header exists, case-insensitive
function hasHeader(reqOpts, header) {
    return 'undefined' !== typeof getHeaderName(reqOpts, header);
}

function toJSONifier(keys) {
    return function () {
        var obj = {};
        var me = this;

        keys.forEach(function (key) {
            if (me[key] && 'function' === typeof me[key].toJSON) {
                obj[key] = me[key].toJSON();
            } else {
                obj[key] = me[key];
            }
        });

        return obj;
    };
}

function setupPipe(resp, opts) {
    // make the response await-able
    var resolve;
    var reject;
    var p = new Promise(function (_resolve, _reject) {
        resolve = _resolve;
        reject = _reject;
    });

    // or an existing write stream
    if ('function' === typeof opts.stream.pipe) {
        if (opts.debug) {
            console.debug('[@root/request] stream piped');
        }
        resp.pipe(opts.stream);
    }
    resp.once('error', function (e) {
        if (opts.debug) {
            console.debug("[@root/request] stream 'error'");
            console.error(e.stack);
        }
        resp.destroy();
        if ('function' === opts.stream.destroy) {
            opts.stream.destroy(e);
        }
        reject(e);
    });
    resp.once('end', function () {
        if (opts.debug) {
            console.debug("[@root/request] stream 'end'");
        }
        if ('function' === opts.stream.destroy) {
            opts.stream.end();
            // this will close the stream (i.e. sync to disk)
            opts.stream.destroy();
        }
    });
    resp.once('close', function () {
        if (opts.debug) {
            console.debug("[@root/request] stream 'close'");
        }
        resolve();
    });
    return p;
}

function handleResponse(resp, opts, cb) {
    // body can be buffer, string, or json
    if (null === opts.encoding) {
        resp._body = [];
    } else {
        resp.body = '';
    }
    resp._bodyLength = 0;
    resp.on('readable', function () {
        var chunk;
        while ((chunk = resp.read())) {
            if ('string' === typeof resp.body) {
                resp.body += chunk.toString(opts.encoding);
            } else {
                resp._body.push(chunk);
                resp._bodyLength += chunk.length;
            }
        }
    });
    resp.once('end', function () {
        if ('string' !== typeof resp.body) {
            if (1 === resp._body.length) {
                resp.body = resp._body[0];
            } else {
                resp.body = Buffer.concat(resp._body, resp._bodyLength);
            }
            resp._body = null;
        }
        if (opts.json && 'string' === typeof resp.body) {
            // TODO I would parse based on Content-Type
            // but request.js doesn't do that.
            try {
                resp.body = JSON.parse(resp.body);
            } catch (e) {
                // ignore
            }
        }

        debug('\n[urequest] resp.toJSON():');
        if (module.exports.debug) {
            debug(resp.toJSON());
        }
        if (opts.debug) {
            console.debug('[@root/request] Response Body:');
            console.debug(resp.body);
        }
        cb(null, resp, resp.body);
    });
}

function setDefaults(defs) {
    defs = defs || {};

    function urequestHelper(opts, cb) {
        debug('\n[urequest] processed options:');
        debug(opts);

        var req;
        var finalOpts = {};

        // allow specifying a file
        if ('string' === typeof opts.stream) {
            if (opts.debug) {
                console.debug('[@root/request] creating file write stream');
            }
            try {
                opts.stream = fs.createWriteStream(opts.stream);
            } catch (e) {
                cb(e);
                return;
            }
        }

        function onResponse(resp) {
            var followRedirect;

            Object.keys(defs).forEach(function (key) {
                if (key in opts && 'undefined' !== typeof opts[key]) {
                    return;
                }
                opts[key] = defs[key];
            });
            followRedirect = opts.followRedirect;

            // copied from WHATWG fetch
            resp.ok = false;
            if (resp.statusCode >= 200 && resp.statusCode < 300) {
                resp.ok = true;
            }

            resp.toJSON = toJSONifier([
                'statusCode',
                'body',
                'ok',
                'headers',
                'request'
            ]);

            resp.request = req;
            resp.request.uri = url.parse(opts.url);
            //resp.request.method = opts.method;
            resp.request.headers = finalOpts.headers;
            resp.request.toJSON = toJSONifier(['uri', 'method', 'headers']);
            if (opts.debug) {
                console.debug('[@root/request] Response Headers:');
                console.debug(resp.toJSON());
            }

            if (
                followRedirect &&
                resp.headers.location &&
                -1 !== [301, 302, 307, 308].indexOf(resp.statusCode)
            ) {
                debug('Following redirect: ' + resp.headers.location);
                if ('GET' !== opts.method && !opts.followAllRedirects) {
                    followRedirect = false;
                }
                if (opts._redirectCount >= opts.maxRedirects) {
                    followRedirect = false;
                }
                if ('function' === opts.followRedirect) {
                    if (!opts.followRedirect(resp)) {
                        followRedirect = false;
                    }
                }
                if (followRedirect) {
                    if (!opts.followOriginalHttpMethod) {
                        opts.method = 'GET';
                        opts.body = null;
                        delete opts.headers[
                            getHeaderName(opts, 'Content-Length')
                        ];
                        delete opts.headers[
                            getHeaderName(opts, 'Transfer-Encoding')
                        ];
                    }
                    if (opts.removeRefererHeader && opts.headers) {
                        delete opts.headers.referer;
                    }
                    // TODO needs baseUrl, maybe test for host / socketPath?
                    opts.url = resp.headers.location;
                    opts.uri = url.parse(opts.url);
                    return urequestHelper(opts, cb);
                }
            }

            if (opts.stream) {
                resp.stream = setupPipe(resp, opts);
                // can be string, buffer, or json... why not an async function too?
                resp.stream.body = async function () {
                    handleResponse(resp, opts, cb);
                    await resp.stream;
                    return resp.body;
                };
                cb(null, resp);
                return;
            }

            handleResponse(resp, opts, cb);
        }

        var _body;
        var MyFormData;
        var form;
        var formHeaders;
        var requester;

        if (opts.body) {
            if (true === opts.json) {
                _body = JSON.stringify(opts.body);
            } else {
                _body = opts.body;
            }
        } else if (opts.json && true !== opts.json) {
            _body = JSON.stringify(opts.json);
        } else if (opts.form) {
            _body = Object.keys(opts.form)
                .filter(function (key) {
                    if ('undefined' !== typeof opts.form[key]) {
                        return true;
                    }
                })
                .map(function (key) {
                    return (
                        encodeURIComponent(key) +
                        '=' +
                        encodeURIComponent(String(opts.form[key]))
                    );
                })
                .join('&');
            opts.headers['Content-Type'] = 'application/x-www-form-urlencoded';
        }
        if ('string' === typeof _body) {
            _body = Buffer.from(_body);
        }

        Object.keys(opts.uri).forEach(function (key) {
            finalOpts[key] = opts.uri[key];
        });

        // A bug should be raised if request does it differently,
        // but I think we're supposed to pass all acceptable options
        // on to the raw http request
        [
            'family',
            'host',
            'localAddress',
            'agent',
            'createConnection',
            'timeout',
            'setHost'
        ].forEach(function (key) {
            finalOpts[key] = opts.uri[key];
        });

        finalOpts.method = opts.method;
        finalOpts.headers = JSON.parse(JSON.stringify(opts.headers));

        var uaHeader = getHeaderName(finalOpts, 'User-Agent') || 'User-Agent';
        // set a default user-agent
        if (!finalOpts.headers[uaHeader]) {
            if (false === finalOpts.headers[uaHeader]) {
                delete finalOpts.headers[uaHeader];
            } else {
                finalOpts.headers[uaHeader] = getUserAgent(opts.userAgent);
            }
        }
        if (_body) {
            // Most APIs expect (or require) Content-Length except in the case of multipart uploads
            // Transfer-Encoding: Chunked (the default) is generally only well-supported downstream
            if (
                'undefined' !== typeof _body.byteLength ||
                'undefined' !== typeof _body.length
            ) {
                finalOpts.headers['Content-Length'] =
                    _body.byteLength || _body.length;
            }
        }
        if (opts.auth) {
            // if opts.uri specifies auth it will be parsed by url.parse and passed directly to the http module
            if ('string' !== typeof opts.auth) {
                opts.auth =
                    (opts.auth.user || opts.auth.username || '') +
                    ':' +
                    (opts.auth.pass || opts.auth.password || '');
            }
            if ('string' === typeof opts.auth) {
                finalOpts.auth = opts.auth;
            }
            if (false === opts.sendImmediately) {
                console.warn(
                    '[Warn] setting `sendImmediately: false` is not yet supported. Please open an issue if this is an important feature that you need.'
                );
            }
            // [request-compat]
            if (opts.auth.bearer) {
                // having a shortcut for base64 encoding makes sense, but this? Eh, whatevs...
                finalOpts.header.Authorization = 'Bearer ' + opts.auth.bearer;
            }
        }
        if (opts.formData) {
            try {
                MyFormData = opts.FormData || require('form-data');
                // potential options https://github.com/felixge/node-combined-stream/blob/master/lib/combined_stream.js#L7-L15
            } catch (e) {
                console.error(
                    'urequest does not include extra dependencies by default'
                );
                console.error(
                    "if you need to use 'form-data' you may install it, like so:"
                );
                console.error('  npm install --save form-data');
                cb(e);
                return;
            }
            try {
                form = new MyFormData();
                Object.keys(opts.formData).forEach(function (key) {
                    function add(key, data, opts) {
                        if (data.value) {
                            opts = data.options;
                            data = data.value;
                        }
                        form.append(key, data, opts);
                    }
                    if (Array.isArray(opts.formData[key])) {
                        opts.formData[key].forEach(function (data) {
                            add(key, data);
                        });
                    } else {
                        add(key, opts.formData[key]);
                    }
                });
            } catch (e) {
                cb(e);
                return;
            }
            formHeaders = form.getHeaders();
            Object.keys(formHeaders).forEach(function (header) {
                finalOpts.headers[header] = formHeaders[header];
            });
        }

        // TODO support unix sockets
        if ('https:' === finalOpts.protocol) {
            // https://nodejs.org/api/https.html#https_https_request_options_callback
            debug('\n[urequest] https.request(opts):');
            debug(finalOpts);
            requester = https;
        } else if ('http:' === finalOpts.protocol) {
            // https://nodejs.org/api/http.html#http_http_request_options_callback
            debug('\n[urequest] http.request(opts):');
            debug(finalOpts);
            requester = http;
        } else {
            cb(new Error("unknown protocol: '" + opts.uri.protocol + "'"));
            return;
        }

        if (form) {
            debug("\n[urequest] '" + finalOpts.method + "' (request) form");
            debug(formHeaders);
            // generally uploads don't use Chunked Encoding (some systems have issues with it)
            // and I don't want to do the work to calculate the content-lengths. This seems to work.
            req = form.submit(finalOpts, function (err, resp) {
                if (err) {
                    cb(err);
                    return;
                }
                onResponse(resp);
                resp.resume();
            });
            //req = requester.request(finalOpts, onResponse);
            //req.on('error', cb);
            //form.pipe(req);
            return;
        }

        if (opts.debug) {
            console.debug('');
            console.debug('[@root/request] Request Options:');
            console.debug(finalOpts);
            if (_body) {
                console.debug('[@root/request] Request Body:');
                console.debug(
                    opts.body || opts.form || opts.formData || opts.json
                );
            }
        }
        req = requester.request(finalOpts, onResponse);
        req.once('error', cb);

        if (_body) {
            debug("\n[urequest] '" + finalOpts.method + "' (request) body");
            debug(_body);
            if ('function' === typeof _body.pipe) {
                // used for chunked encoding
                _body.pipe(req);
                _body.once('error', function (err) {
                    // https://nodejs.org/api/stream.html#stream_readable_pipe_destination_options
                    // if the Readable stream emits an error during processing,
                    // the Writable destination is not closed automatically
                    _body.destroy();
                    req.destroy(err);
                });
            } else {
                // used for known content-length
                req.end(_body);
            }
        } else {
            req.end();
        }
    }

    function parseUrl(str) {
        var obj = url.parse(str);
        var paths;
        if ('unix' !== (obj.hostname || obj.host || '').toLowerCase()) {
            return obj;
        }

        obj.href = null;
        obj.hostname = obj.host = null;

        paths = (obj.pathname || obj.path || '').split(':');

        obj.socketPath = paths.shift();
        obj.pathname = obj.path = paths.join(':');

        return obj;
    }

    function urequest(opts, cb) {
        debug('\n[urequest] received options:');
        debug(opts);
        var reqOpts = {};
        // request.js behavior:
        // encoding: null + json ? unknown
        // json => attempt to parse, fail silently
        // encoding => buffer.toString(encoding)
        // null === encoding => Buffer.concat(buffers)
        if ('string' === typeof opts) {
            opts = { url: opts };
        }

        module.exports._keys.forEach(function (key) {
            if (key in opts && 'undefined' !== typeof opts[key]) {
                reqOpts[key] = opts[key];
            } else if (key in defs) {
                reqOpts[key] = defs[key];
            }
        });

        // TODO url.resolve(defs.baseUrl, opts.url);
        if ('string' === typeof opts.url || 'string' === typeof opts.uri) {
            if ('string' === typeof opts.url) {
                reqOpts.url = opts.url;
                reqOpts.uri = parseUrl(opts.url);
            } else if ('string' === typeof opts.uri) {
                reqOpts.url = opts.uri;
                reqOpts.uri = parseUrl(opts.uri);
            }
        } else {
            if ('object' === typeof opts.uri) {
                reqOpts.url = url.format(opts.uri);
                reqOpts.uri = opts.uri;
                //reqOpts.uri = url.parse(reqOpts.uri);
            } else if ('object' === typeof opts.url) {
                reqOpts.url = url.format(opts.url);
                reqOpts.uri = opts.url;
                //reqOpts.uri = url.parse(reqOpts.url);
            }
        }

        if (
            opts.body ||
            (opts.json && true !== opts.json) ||
            opts.form ||
            opts.formData
        ) {
            // TODO this is probably a deviation from request's API
            // need to check and probably eliminate it
            reqOpts.method = (reqOpts.method || 'POST').toUpperCase();
        } else {
            reqOpts.method = (reqOpts.method || 'GET').toUpperCase();
        }
        if (!reqOpts.headers) {
            reqOpts.headers = {};
        }

        // crazy case for easier testing
        if (!hasHeader(reqOpts, 'CoNTeNT-TyPe')) {
            if (
                (true === reqOpts.json && reqOpts.body) ||
                (true !== reqOpts.json && reqOpts.json)
            ) {
                reqOpts.headers['Content-Type'] = 'application/json';
            }
        }

        if (opts.debug) {
            reqOpts.debug = opts.debug;
        }
        return urequestHelper(reqOpts, cb);
    }

    function smartPromisify(fn) {
        return function (opts) {
            var cb;
            if ('function' === typeof arguments[1]) {
                cb = Array.prototype.slice.call(arguments)[1];
                return fn(opts, cb);
            }
            return new Promise(function (resolve, reject) {
                fn(opts, function (err, resp) {
                    if (err) {
                        err._response = resp;
                        reject(err);
                        return;
                    }
                    resolve(resp);
                });
            });
        };
    }

    var smartUrequest = smartPromisify(urequest);

    smartUrequest.defaults = function (_defs) {
        _defs = mergeOrDelete(defs, _defs);
        return setDefaults(_defs);
    };
    ['get', 'put', 'post', 'patch', 'delete', 'head', 'options'].forEach(
        function (method) {
            smartUrequest[method] = smartPromisify(function (obj, cb) {
                if ('string' === typeof obj) {
                    obj = { url: obj };
                }
                obj.method = method.toUpperCase();
                urequest(obj, cb);
            });
        }
    );
    smartUrequest.del = urequest.delete;

    return smartUrequest;
}

var nodeUa =
    '@root+request/' +
    pkg.version +
    ' ' +
    process.release.name +
    '/' +
    process.version +
    ' ' +
    os.platform() +
    '/' +
    os.release() +
    ' ' +
    os.type() +
    '/' +
    process.arch;
function getUserAgent(additional) {
    // See https://tools.ietf.org/html/rfc8555#section-6.1
    // And https://tools.ietf.org/html/rfc7231#section-5.5.3
    // And https://community.letsencrypt.org/t/user-agent-flag-explained/3843/2

    var ua = nodeUa;
    if (additional) {
        ua = additional + ' ' + ua;
    }

    return ua;
}

var _defaults = {
    sendImmediately: true,
    method: '',
    headers: {},
    useQuerystring: false,
    followRedirect: true,
    followAllRedirects: false,
    followOriginalHttpMethod: false,
    maxRedirects: 10,
    removeRefererHeader: false,
    // encoding: undefined,
    // stream: false, // TODO allow a stream?
    gzip: false
    //, body: undefined
    //, json: undefined
};
module.exports = setDefaults(_defaults);

module.exports._keys = Object.keys(_defaults).concat([
    'encoding',
    'stream',
    'body',
    'json',
    'form',
    'auth',
    'formData',
    'FormData',
    'userAgent' // non-standard for request.js
]);
module.exports.debug =
    -1 !== (process.env.NODE_DEBUG || '').split(/\s+/g).indexOf('urequest');

debug('DEBUG ON for urequest');
